Add card buying

oliverlee opened the pull request.

oliverlee committed:
Add player method to buy cards

Use player turn phase and resources determine if a card can be
purchased. This does not decrease the card from the supply.

oliverlee committed:
Decrease cards from supply when bought

Add supply reference to players. Validate card is in base cards or
kingdom cards and that the supply count is non-zero when buying cards.

oliverlee committed:
Refactor play_card to use CardKind

Use CardKind instead of an index for Player::play_card. Replace
player::Error::{InvalidCardIndex, InvalidSupplyChoice} with
player::Error::InvalidCardChoice. Perform cleanup when ending turn.

andebjor commented on src/dominion/player.rs:
```diff
@@ -191,10 +234,81 @@ impl Player {
+        let r = p.buy_card(&CardKind::Copper);
+
+        assert!(r.is_ok());
+        assert_eq!(p.discard_pile[0], &CardKind::Copper);
```
Assert that the supply of coppers have decreased by one?

oliverlee replied:
done

andebjor commented on src/dominion/player.rs:
```diff
@@ -191,10 +234,81 @@ impl Player {
+        let r = p.buy_card(&CardKind::Market);
+
+        assert!(r.is_ok());
+        assert_eq!(p.discard_pile[0], &CardKind::Market);
```
Assert that wealth is now 0?

andebjor commented on src/dominion/player.rs:
```diff
@@ -136,6 +142,43 @@ impl Player {
         }
     }
 
+    pub fn buy_card(&mut self, c: &'static CardKind) -> Result<()> {
```
I'm not a fan of single letter arguments, such as `c` :)

oliverlee commented:
#![feature(vec_remove_item)] only works on nightly. Should we drop support for stable and beta?

mickvangelderen commented:
> #![feature(vec_remove_item)] only works on nightly. Should we drop support for stable and beta?

For a toy project yes, we can worry about running this in stable later (read: probably never).

mickvangelderen reviewed (commented):
I am concerned about the general architecture. The simulation we are running concerns all of the game state (currently kept in Arena). Player hands are just a part of that. I think it is better to do all of the mutations in a function which has access to the entire state. This will remove the need for reference counted supply pointers (and probably more in the future if this design is continued).

I am okay with going forward with this architecture for the sake of learning though.

oliverlee committed:
Disable rust stable, beta toolchains in Travis

Due to use of feature vec_remove_item, stable and beta toolchains are
not tested.

See: https://github.com/rust-lang/rust/issues/40062

oliverlee commented:
> I am concerned about the general architecture. The simulation we are running concerns all of the game state (currently kept in Arena). Player hands are just a part of that. I think it is better to do all of the mutations in a function which has access to the entire state. This will remove the need for reference counted supply pointers (and probably more in the future if this design is continued).
> 
> I am okay with going forward with this architecture for the sake of learning though.

This is mostly a learning exercise but I think we should change the architecture if we can come up with a better idea. Resolving a stack of actions as a result of playing a card will probably be challenging with this approach.
